Mousemove event should be fired when the mouse stays in the document but there is no element

Ahmad-S792 opened the pull request:
#### 7a8eec33525ba65a414942b713e7454fb3705753
<pre>
Mousemove event should be fired when the mouse stays in the document but there is no element

<a href="https://bugs.webkit.org/show_bug.cgi?id=266932">https://bugs.webkit.org/show_bug.cgi?id=266932</a>
<a href="https://rdar.apple.com/120551245">rdar://120551245</a>

Reviewed by Abrar Rahman Protyasha.

Merge: <a href="https://github.com/chromium/chromium/commit/20a8667a4e0d5e28d4f301ad5c66a0fa3b540ca6">https://github.com/chromium/chromium/commit/20a8667a4e0d5e28d4f301ad5c66a0fa3b540ca6</a>

This pathc fixes an issue where we don&apos;t fire `mousemove` events in the document,
where there is no element. It is fixed by improving our hittesting code and leads
to match other browser engine (i.e., Blink / Chromium).

* LayoutTests/fast/events/mousemove-events-within-no-element.html:
* LayoutTests/fast/events/resources/mouse-events-within-no-element-iframe.html:
* LayoutTests/fast/events/mousemove-events-within-no-element-expected.txt:
* LayoutTests/platform/ios/TestExpectations: Add iOS Specific expectation due to lack of &apos;mouse events&apos;
* Source/WebCore/rendering/RenderLayer.cpp:
(WebCore::RenderLayer::hitTest):

Canonical link: <a href="https://commits.webkit.org/296341@main">https://commits.webkit.org/296341@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/207b66d7f2148e42caf2bc542f013d78273042b2

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/108243 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/27905 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/18326 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/113452 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/58704 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/110206 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/28602 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/36456 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/82190 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/58704 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/111191 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/22666 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/97506 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/62622 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/22079 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/15642 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/58186 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/92032 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/15698 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/116576 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/35305 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/25993 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/91214 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/35678 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/93781 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/91010 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/23191 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/35901 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/13666 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/31073 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/35205 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/40753 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/34934 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/38288 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/36603 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

aprotyas reviewed (commented):
Small typo in the commit message:

> This ~pathc~patch fixes

aprotyas reviewed (commented):
Another nit about the commit message, wouldn't the `Source` entry be at the end?

```diff
 * LayoutTests/platform/ios/TestExpectations:
-* Source/WebCore/rendering/RenderLayer.cpp:
-(WebCore::RenderLayer::hitTest):
 * LayoutTests/fast/events/mousemove-events-within-no-element.html:
 * LayoutTests/fast/events/resources/mouse-events-within-no-element-iframe.html:
 * LayoutTests/fast/events/mousemove-events-within-no-element-expected.txt:
+* Source/WebCore/rendering/RenderLayer.cpp:
+(WebCore::RenderLayer::hitTest):
```

aprotyas commented on Source/WebCore/rendering/RenderLayer.cpp:
```diff
@@ -4502,7 +4502,10 @@ bool RenderLayer::hitTest(const HitTestRequest& request, const HitTestLocation&
+        // In addtion, it is possible for the mouse to stay in the document but there is no element.
+        // At that time, the events of the mouse should be fired.
+        LayoutPoint hitPoint = hitTestLocation.point();
+        if (!request.isChildFrameHitTest() && ((request.active() || request.release()) || (request.move() && hitTestArea.contains(hitPoint.x(), hitPoint.y()))) && isRenderViewLayer()) {
```
This condition is getting long, do you mind breaking out the new logic into a separate predicate? Also, `LayoutRect::contains(const LayoutPoint&)` is a thing, so your `contains()` call can be made more concise.

```suggestion
        bool moveRequestIsOverDocument = request.move() && hitTestArea.contains(hitPoint);
        if (!request.isChildFrameHitTest() && (request.active() || request.release() || moveRequestIsOverDocument) && isRenderViewLayer()) {
```

Ahmad-S792 replied:
Done.

aprotyas reviewed (approved):
r+ with my suggestion

Ahmad-S792 committed:
Mousemove event should be fired when the mouse stays in the document but there is no element

https://bugs.webkit.org/show_bug.cgi?id=266932
rdar://120551245

Reviewed by Abrar Rahman Protyasha.

Merge: https://github.com/chromium/chromium/commit/20a8667a4e0d5e28d4f301ad5c66a0fa3b540ca6

This pathc fixes an issue where we don't fire `mousemove` events in the document,
where there is no element. It is fixed by improving our hittesting code and leads
to match other browser engine (i.e., Blink / Chromium).

* LayoutTests/fast/events/mousemove-events-within-no-element.html:
* LayoutTests/fast/events/resources/mouse-events-within-no-element-iframe.html:
* LayoutTests/fast/events/mousemove-events-within-no-element-expected.txt:
* LayoutTests/platform/ios/TestExpectations: Add iOS Specific expectation due to lack of 'mouse events'
* Source/WebCore/rendering/RenderLayer.cpp:
(WebCore::RenderLayer::hitTest):

Canonical link: https://commits.webkit.org/296341@main
